File: advent2018_23.py
import math

bots = []
f = open('input.txt','r')
for line in f.readlines():
	line = line.split(', r=')
	bots.append((map(int,line[0][5:-1].split(',')), int(line[1])))
f.close()

mindist = float('inf')
minx, maxx, miny, maxy, minz, maxz = 0, 0, 0, 0, 0, 0

for b in bots:
	minx = min(minx,b[0][0])
	maxx = max(maxx,b[0][0])
	miny = min(miny,b[0][1])
	maxy = max(maxy,b[0][1])
	minz = min(minz,b[0][2])
	maxz = max(maxz,b[0][2])
	mindist = min(mindist,b[1])

def inrange(pos,bot):
	return abs(pos[0]-bot[0][0])+abs(pos[1]-bot[0][1])+abs(pos[2]-bot[0][2]) <= bot[1]

def ninrange(pos):
	c = 0
	for b in bots:
		if inrange(pos,b): c += 1
	return c

while mindist > 0:
	loc = {}
	print mindist
	offset = mindist*9/11

	for x in xrange(minx-offset,maxx+1,mindist):
		for y in xrange(miny-offset,maxy+1,mindist):
			for z in xrange(minz-offset,maxz+1,mindist):
				loc[(x,y,z)] = ninrange((x,y,z))

	m = max(loc.values())

	opt = []
	for k in loc.keys():
		if loc[k] == m:
			opt.append(k)
	nearest = sorted(opt,key = sum)[0]

	mindist /= 2
	minx = (minx+nearest[0])/2
	maxx = (maxx+nearest[0])/2
	miny = (miny+nearest[1])/2
	maxy = (maxy+nearest[1])/2
	minz = (minz+nearest[2])/2
	maxz = (maxz+nearest[2])/2

	if mindist == 0: 
		break

print nearest,sum(nearest),m

File: Advent2018_23.java
import java.util.*;
import java.io.*;

class Bot implements Comparable<Bot> {

	public long x,y,z,r;

	public Bot(long x, long y, long z, long r) {
		this.x = x;
		this.y = y;
		this.z = z;
		this.r = r;
	}

	public boolean inRange(long x, long y, long z) {
		return Math.abs(x-this.x)+Math.abs(y-this.y)+Math.abs(z-this.z) <= r;
	}

	public int compareTo(Bot other) {
		return (this.x+this.y+this.z) > (other.x+other.y+other.z) ? 1 : -1;
	}

}

public class Advent2018_23 {

	public static void main(String[] args) throws FileNotFoundException {
		Scanner s = new Scanner(new File("input.txt"));
		ArrayList<Bot> bots = new ArrayList<Bot>();
		while (s.hasNext()) {
			String line = s.nextLine();
			bots.add(new Bot(Integer.parseInt(line.split("<|,|=|>")[2]),Integer.parseInt(line.split("<|,|=|>")[3]),Integer.parseInt(line.split("<|,|=|>")[4]),Integer.parseInt(line.split("<|,|=|>")[7]))); // stolen
		}
		// we now have a list of bots
		long ix,ax,iy,ay,iz,az,window;
		ix = -200000000;
		ax = 200000000;
		iy = -200000000;
		ay = 200000000;
		iz = -200000000;
		az = 200000000;
		window = 4000000;
		Bot maxbot = new Bot(0, 0, 0, 0);
		int maxn = 0;

		while (window > 0) {

			maxn = 0;
			maxbot = new Bot(ax, ay, az, 0);

			for (long x = ix; x < ax; x += window) {
				for (long y = iy; y < ay; y += window) {
					for (long z = iz; z < az; z += window) {
						int k = nInRange(x, y, z, bots);
						if (k > maxn || k == maxn && maxbot.compareTo(new Bot(x, y, z, 0)) > 0) {
							maxn = k;
							maxbot = new Bot(x, y, z, 0);
						}
					}
				}
			}
			System.out.println(maxbot.x+", "+maxbot.y+", "+maxbot.z+"; n = "+maxn);
			
			window /= 2;
			ix = (ix+maxbot.x)/2;
			ax = (ax+maxbot.x)/2;
			iy = (iy+maxbot.y)/2;
			ay = (ay+maxbot.y)/2;
			iz = (iz+maxbot.z)/2;
			az = (az+maxbot.z)/2;

		}

		System.out.println("distance = "+(maxbot.x+maxbot.y+maxbot.z));
	}

	public static int nInRange(long x, long y, long z, List<Bot> bots) {
		int c = 0;
		for (Bot b : bots) {
			if (b.inRange(x, y, z)) c++;
		}
		return c;
	}

}